plugins/org.eclipse.fordiac.ide.typemanagement/src/org/eclipse/fordiac/ide/typemanagement/refactoring/copy/CopyTypeChange.java

```
@@ -0,0 +1,103 @@
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.fordiac.ide.model.helpers.PackageNameHelper;
+import org.eclipse.fordiac.ide.model.libraryElement.LibraryElement;
+import org.eclipse.fordiac.ide.model.libraryElement.LibraryElementFactory;
+import org.eclipse.fordiac.ide.typemanagement.Messages;
+import org.eclipse.fordiac.ide.ui.FordiacLogHelper;
+import org.eclipse.ltk.core.refactoring.Change;
+import org.eclipse.ltk.core.refactoring.RefactoringStatus;
+
+public class CopyTypeChange extends Change {
+
+	private final String name;
+	private final IFile origin;
+	private final URI destination;
+	private final String newPackageName;
+	private String oldPackageName;
+
+	protected CopyTypeChange(final String newPackageName, final String name, final IFile origin,
+			final URI destination) {
+		this.newPackageName = newPackageName;
+		this.name = name;
+		this.origin = origin;
+		this.destination = destination;
+	}
+
+	@Override
+	public void initializeValidationData(final IProgressMonitor pm) {
+		oldPackageName = PackageNameHelper.getPackageNameFromFile(origin);
+	}
+
+	@Override
+	public RefactoringStatus isValid(final IProgressMonitor pm) throws CoreException, OperationCanceledException {
+		final RefactoringStatus status = new RefactoringStatus();
+		if (Objects.equals(oldPackageName, newPackageName)) {
+			status.addWarning(Messages.MoveTypeToPackage_PackageNameIsTheSame);
+		}
+		return status;
+	}
+
+	@Override
+	public String getName() {
+		return name;
+	}
+
+	@Override
+	public Change perform(final IProgressMonitor pm) throws CoreException {
+		final var resource = CopyTypeParticipant.getResource(destination);
+		if (resource.isEmpty()) {
+			return null;
```

Please throw an appropriate exception if the resource cannot be loaded to abort the refactoring, e.g. https://github.com/eclipse-4diac/4diac-ide/blob/27baaf1560b8af1e587c4398f2ff7da87f6d2dc8/plugins/org.eclipse.fordiac.ide.typemanagement/src/org/eclipse/fordiac/ide/typemanagement/refactoring/AbstractCommandChange.java#L159-L163